transports/quic/src/transport.rs

```diff
@@ -136,75 +166,166 @@ impl Transport for QuicTransport {
+    /// time as interfaces become available or unavailable.
+    in_addr: InAddr,
+
+    /// Set to `Some` if this [`Listener`] should close.
+    /// Optionally contains a [`TransportEvent::ListenerClosed`] that should be
+    /// reported before the listener's stream is terminated.
+    report_closed: Option<Option<<Self as Stream>::Item>>,
+}
+
+impl Listener {
+    fn new(
+        listener_id: ListenerId,
+        socket_addr: SocketAddr,
+        config: Config,
+    ) -> Result<Self, Error> {
+        let in_addr = InAddr::new(socket_addr.ip());
+        let (endpoint, new_connections_rx) = Endpoint::new_bidirectional(config, socket_addr)?;
+        Ok(Listener {
+            endpoint,
+            listener_id,
+            new_connections_rx,
+            in_addr,
+            report_closed: None,
+        })
+    }
+
+    /// Report the listener as closed in a [`TransportEvent::ListenerClosed`] and
+    /// terminate the stream.
+    fn close(&mut self, reason: Result<(), Error>) {
+        match self.report_closed {
+            Some(_) => tracing::debug!("Listener was already closed."),
+            None => {
+                // Report the listener event as closed.
+                let _ = self
+                    .report_closed
+                    .insert(Some(TransportEvent::ListenerClosed {
+                        listener_id: self.listener_id,
+                        reason,
+                    }));
+            }
+        }
+    }
+
+    /// Poll for a next If Event.
+    fn poll_if_addr(&mut self, cx: &mut Context<'_>) -> Option<<Self as Stream>::Item> {
+        match self.in_addr.poll_next_unpin(cx) {
             Poll::Ready(mut item) => {
                 if let Some(item) = item.take() {
                     // Consume all events for up/down interface changes.
                     match item {
                         Ok(IfEvent::Up(inet)) => {
                             let ip = inet.addr();
-                            if endpoint.local_addr.is_ipv4() == ip.is_ipv4() {
-                                let socket_addr = SocketAddr::new(ip, endpoint.local_addr.port());
+                            if self.endpoint.socket_addr().is_ipv4() == ip.is_ipv4() {
+                                let socket_addr =
+                                    SocketAddr::new(ip, self.endpoint.socket_addr().port());
                                 let ma = socketaddr_to_multiaddr(&socket_addr);
                                 tracing::debug!("New listen address: {}", ma);
-                                return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(ma))));
+                                Some(TransportEvent::NewAddress {
+                                    listener_id: self.listener_id,
+                                    listen_addr: ma,
+                                })
+                            } else {
+                                self.poll_if_addr(cx)
```

Comment:
@kpp:
> Are we okay with an infinite recursion in case there are infinite many IPv6 events when we are bound to IPv4?

Reply:
That can only happen if the `IfWatcher` will always return `Poll::Ready(ipv6_event)` if we poll it, which would be a problematic bug either way, no?

I'd like to catch the case here that the `IfWatcher` first reports an IPv6 address and directly afterwards an Ipv4 one. In that situation we can not just return `Poll::Pending` here but instead have to solve it by either a) wrapping the whole block in a loop and `continue` if an Ipv6 address is reported, or b) do this recursion. In both cases the behaviour that you described would be a problem. No sure if this can be solved any other way?

Reply:
> which would be a problematic bug either way, no?

No. You can delete ipv4 from an interface, so you will always get ipv6 events from IfWatcher. An infinite loop will crash the code, a missing ipv4 addr will not.

Reply:
> > which would be a problematic bug either way, no?
> 
> No. You can delete ipv4 from an interface, so you will always get ipv6 events from IfWatcher. An infinite loop will crash the code, a missing ipv4 addr will not.

I am not sure I am following @kpp. Are you saying that if-watch could potentially return ipv6 events non-stop, i.e. without a single `Poll::Pending` in between?

> a) wrapping the whole block in a loop and `continue` if an Ipv6 address is reported,

Note that Rust does not have tail recursion, thus large recursive nesting could potentially be problematic in regards to the stack size. I doubt this would be a problem here, in other words I expect this to only recurse little.

From a different perspective, I don't think recursion is idiomatic Rust. Thus I would favor a `loop`. What do you think @elenaf9?

See also https://github.com/libp2p/rust-libp2p/pull/2638#discussion_r873945621

Reply:
> From a different perspective, I don't think recursion is idiomatic Rust. Thus I would favor a loop. What do you think @elenaf9?

Done by @kpp in elenaf9#5 / 574d534.